internal/pkg/sourceinfer/analyzer.go

```diff
@@ -0,0 +1,285 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package infer defines an analyzer that identifies Sources that either
+// 1. are defined from a Source type, or
+// 2. have a Source type as a field
+package infer
+
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+	"go/types"
+	"reflect"
+
+	"github.com/google/go-flow-levee/internal/pkg/config"
+	"github.com/google/go-flow-levee/internal/pkg/utils"
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/passes/buildssa"
+	"golang.org/x/tools/go/analysis/passes/inspect"
+	"golang.org/x/tools/go/ast/inspector"
+)
+
+// ResultType is a set of types.Object that are inferred Sources.
+type ResultType map[types.Object]bool
+
+type inferredSource struct{}
+
+func (i inferredSource) AFact() {}
+
+func (i inferredSource) String() string {
+	return "inferred source"
+}
+
+var Analyzer = &analysis.Analyzer{
+	Name: "sourceinfer",
+	Doc: `This analyzer infers source types from typedefs and fields.
+
+Suppose Foo has been configured to be a source type.
+
+Then, the following type definition would cause Bar to be identified
+as a source, because its underlying type is Foo:
+type Bar Foo
+Indeed, if Foo contains sensitive data, then so does Bar, because they have
+the same fields.
+
+Similarly, the following type definition would cause Baz to be identified
+as a source, because it holds a field of type Foo:
+type Baz struct {
+  f Foo
+}
+Indeed, if Foo contains sensitive data, and Bar contains Foo, Bar also
+contains that data, via Foo.
+
+The analysis is recursive: types identified as sources during analysis
```

Comment:
Nit: given the graph you build out of types, this algorithm is not recursive, is it?  You maybe meant it as the closure suggested above?

Reply:
In my understanding, the *analysis* is recursive, since inferring that a type is a source leads the analysis to try to infer further sources from that type. Actually, in this case I guess *iterative* could also work...

Whether the *algorithm* itself is recursive is debatable, since it's basically DFS. There is a recursive DFS in there (toposort) and a non-recursive DFS as well (the sourceyness propagation proper).

What I'm hearing is that the term "recursive" here isn't very helpful, so I've removed that part altogether. I think the resulting description is clearer.